Add FileOpenedEvent

dstaple opened the pull request:
As per https://github.com/gorakhargosh/watchdog/issues/901 , this adds support to detect file open events on supported OSes.

I have tested this in Linux using Python 3.7 and Python 3.11. I could use help testing it on other OSes, in particular, changes may be needed in order to fail gracefully on OSes where FileOpenEvents are not supported.

BoboTiG commented:
Thanks a lot @dstaple 🍾 

Would you mind adding a line + your GitHub nickname in the changelog?

BoboTiG commented:
About kqueue, do you think you can add the new event support too?

dstaple commented:
Thanks @BoboTiG ! I updated the changelog and fixed the linting error.

Are you OK with me leaving the kqueue observer out of this PR?

kqueue is also missing FileClosedEvent, and with the way it's implemented in terms of DirectorySnapshot, it's not immediately obvious to me how this will work.

BoboTiG commented:
Yes, that's already awesome!
Thanks @dstaple :)

Xyene commented:
Just a heads up, this change resulted in a performance regression in downstream software that's been happily using watchdog for ~10 years (https://github.com/DMOJ/judge-server/issues/1129), which cares about everything but open/close events (where it is being fixed by filtering out the open/close events). It is a _little_ surprising that possibly very-noisy open events are being tracked by default, but I don't feel strongly about this.

Thought it was worth mentioning in case we were not the only ones who will run into this, as downstream projects upgrade to newer versions of watchdog 🙂
